src/ripple/protocol/impl/Feature.cpp

```diff
@@ -25,41 +26,180 @@
 
 namespace ripple {
 
-//------------------------------------------------------------------------------
+enum class Supported : bool { no = false, yes };
 
-constexpr char const* const detail::FeatureCollections::featureNames[];
+namespace detail {
+// *NOTE*
+//
+// Features, or Amendments as they are called elsewhere, are enabled on the
+// network at some specific time based on Validator voting.  Features are
+// enabled using run-time conditionals based on the state of the amendment.
+// There is value in retaining that conditional code for some time after
+// the amendment is enabled to make it simple to replay old transactions.
+// However, once an Amendment has been enabled for, say, more than two years
+// then retaining that conditional code has less value since it is
+// uncommon to replay such old transactions.
+//
+// Starting in January of 2020 Amendment conditionals from before January
+// 2018 are being removed.  So replaying any ledger from before January
+// 2018 needs to happen on an older version of the server code.  There's
+// a log message in Application.cpp that warns about replaying old ledgers.
+//
+// At some point in the future someone may wish to remove Amendment
+// conditional code for Amendments that were enabled after January 2018.
+// When that happens then the log message in Application.cpp should be
+// updated.
 
-detail::FeatureCollections::FeatureCollections()
+class FeatureCollections
 {
-    features.reserve(numFeatures());
-    featureToIndex.reserve(numFeatures());
-    nameToFeature.reserve(numFeatures());
+    struct Feature
+    {
+        std::string name;
+        uint256 feature;
+        std::size_t index;
+
+        explicit Feature() = default;
+        explicit Feature(
+            std::string const& name_,
+            uint256 const& feature_,
+            std::size_t index_)
+            : name(name_), feature(feature_), index(index_)
+        {
+        }
+    };
+    std::vector<Feature> features;
```

Comment:
Consider using `boost::multi_index::multi_index_container`. This way `features`, `featureToIndex`, and `nameToFeature` can be consolidated into one container. Also `index` can be removed from `Feature`. Something like this:

     struct ByName{};
     struct ByFeature{};
     struct Feature {
          std::string name;
          uint256 feature;
     };
     using namespace boost::multi_index;
 
     multi_index_container<
         Feature,
         indexed_by<
             random_access<>,
             hashed_unique<tag<ByFeature>, member<Feature, uint256, &Feature::feature>>,
             hashed_unique<tag<ByName>, member<Feature, std::string, &Feature::name>>
         >
     > features;
     // get by feature
     const auto& feature_index = features.get<1>();
     auto feature_it = feature_index.find(101);
     // get by name
     const auto& name_index = features.get<2>();
     auto name_it = name_index.find("featureName");
     // get by index
     const auto& sequence = features.get<0>();
     auto f = sequence[0];
     // get index from name
     auto it_to = sequence.iterator_to(*name_it);
     auto index = it_to - sequence.begin();

Reply:
Good idea. I had considered using `boost::intrusive::multiset`, but wasn't going to bother unless this got some traction, then forgot about it. I think the `multi_index_container` works even better in this case. Thanks!

Reply:
Nice!  I've considered using `multi_index_container` to replace uses of the aged containers.  It would be good to get some experience with that container to see how we like it.